### ISSUE TITLE
Question on activation regularization

### ISSUE THREAD

**Issue by a-maci** [ppwwyyxx/tensorpack#72]
1. How do we regularize activations (output of conv or fc layers) using tensorpack?

Say I have this code snippet:
```
 ## l0 is an input tensor
 l1 = Conv2D('conv1', l0, ch_out, 1, nl=BNReLU)
 l2 = Conv2D('conv2', l1, ch_out, 3, stride=stride, nl=BNReLU)
 l3 = Conv2D('conv3', l2, ch_out * 4, 1)
```
Goal is to regularize l1, l2 and l3 (the activation output from the convolutional layers).
Here is what my current thought is ...

```
 ## l0 is an input tensor
 l1 = Conv2D('conv1', l0, ch_out, 1, nl=BNReLU)
 l1r = regularize_cost('??', tf.nn.l2_loss) # not sure what ?? should be (l1.* ?)
 l2 = Conv2D('conv2', l1r, ch_out, 3, stride=stride, nl=BNReLU)
 l2r = regularize_cost('??', tf.nn.l2_loss) # not sure what ?? should be (l2.* ?)
 l3 = Conv2D('conv3', l2r, ch_out * 4, 1)
 l3r = regularize_cost('??', tf.nn.l2_loss) # not sure what ?? should be (l3.* ?)

## add l1r, l2r and l3r to the cost function
  ##pass
```

2. TF allows you to add a regularizer to a variable at define time (tf.get_variable). Why do the examples not follow this approach? I saw some examples explicitly adding a regularization statement, (e.g. mnist example does wd_cost = tf.mul(1e-5, regularize_cost('fc.*/W', tf.nn.l2_loss), name='regularize_loss') ).

3. What is a good way to add regularizations to the activations (outputs) using the LinearWrap syntax?

**Comment by a-maci** [comment]
Actually is there anything like activation regularization? :)
Can one regularize the activations?

**Comment by ppwwyyxx** [comment]
1. `regularize_cost` is designed only for regularizing variables, so it doesn't work for activations.
You can use `l1r = 0.001 * tf.nn.l2_loss(l1)`, and then add `l1r` to cost.

2. `get_variable` with regularizer option would add the loss to a collection. Then you would still need to get that collection and add it to the total cost. 
Allowing this option would require most layers in tensorpack to have extra options (W_regularize, b_regularize, etc) and this doesn't make things much easier. I think regularize_cost with regular expression is OK.

3. This is a lot of things that cannot be done with LinearWrap. It's just a syntax sugar to use when your model is very simple. And regularizing activations is not a common things people do, so there is no functionality specifically designed for that.
